### PR TITLE
RTEMIS-154: Fixed the Admin Section BE Issues.

### PR DESCRIPTION
**Pull request by RohanSingh-inno**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-154

---
### What the PR does
`(Explain fixes, changes and impacts)`
-  Modified Menus for State and district.
- Added Custom Access Checks.
- Modified Configs for Administer User By Role.
- Modified the Action link on People Listing Page.

### CODE REVIEW

File: modules/rte_mis_state/rte_mis_state.module [hunk 1]
```diff
@@ -209,6 +234,44 @@ function rte_mis_state_form_alter(&$form, FormStateInterface $form_state, $form_
...
+          ];
+        }
+        elseif ($role_param == 'block_admin') {
+          $form['role']['#options'] = [
+            'All' => 'Any',
+            'school' => 'School',
+            'school_admin' => 'School Admin',
+            'block_admin' => 'Block Admin',
+            'district_admin' => 'District Admin',
+          ];
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:260]: @RohanSingh-inno Can you please check if **school**, and **school admin** should be created via `Add user` operation, According to workflow, the school should register the account from the school registration form and their account should not be created by admins.
CC - @kunal-sin

File: modules/rte_mis_state/rte_mis_state.module [hunk 2]
```diff
@@ -445,6 +508,129 @@ function rte_mis_state_views_query_alter(ViewExecutable $view, QueryPluginBase $
...
+          // Load the tree if location based in block.
+          if (!empty($block_id)) {
+            $location_tree = \Drupal::entityTypeManager()->getStorage('taxonomy_term')->loadTree('location', $block_id) ?? [];
+          }
+          if (!empty($location_tree)) {
+            foreach ($location_tree as $value) {
+              $id[] = $value->tid;
+            }
+          }
+        }
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:616]: This line of code is being used in the previous conditions as well. Can we create a common function for this?

File: modules/rte_mis_state/rte_mis_state.module (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_state/rte_mis_state.module:260]: This code gives the option to the state admin to list down the users with school and school admin role. While creating a new user we are still not giving the access to state admin to create any **school/school_admin** account.

File: modules/rte_mis_state/rte_mis_state.module [hunk 3]
```diff
@@ -445,6 +508,129 @@ function rte_mis_state_views_query_alter(ViewExecutable $view, QueryPluginBase $
...
+            $value = implode(',', $field_location_arr);
+          }
+          $individual = 'individual';
+          // Adding an extra condition to the existing conditions.
+          $query->addWhereExpression(1, "taxonomy_term__field_location.field_location_target_id IN ($value)");
+          $query->addWhereExpression(1, "taxonomy_term__field_upload_type.field_upload_type_value = '$individual'");
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (RohanSingh-inno) [reply modules/rte_mis_state/rte_mis_state.module]: Yes, we can remove this.

File: modules/rte_mis_state/rte_mis_state.module (hunk 2 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_state/rte_mis_state.module]: Ok

File: modules/rte_mis_state/rte_mis_state.module (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_state/rte_mis_state.module]: We have modified the permissions. Now State cannot edit **school/ school_admin** accounts.

File: modules/rte_mis_state/rte_mis_state.module [hunk 4]
```diff
@@ -685,3 +769,56 @@ function rte_mis_state_mail($key, &$message, $params) {
...
+}
+
+/**
+ * Function to find the childs of a location of the User.
+ */
+function rte_mis_state_get_location_child($current_user) {
+  $id = [0];
+  // Get the location details from user entity.
+  $selected_location = $current_user->get('field_location_details')->entity ?? NULL;
+  if ($selected_location instanceof TermInterface) {
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:809]: Use the parameter type of user interface

File: modules/rte_mis_state/rte_mis_state.services.yml [hunk 5]
```diff
@@ -9,3 +9,9 @@ services:
     class: Drupal\rte_mis_state\Routing\RouteSubscriber
     tags:
       - { name: event_subscriber }
+
+  rte_mis_state.user_edit_check:
+    class: Drupal\rte_mis_state\Access\UserEditAccessCheck
+    arguments: ['@entity_type.manager', '@current_user']
+    tags:
+      - { name: access_check, applies_to: _user_edit_check }
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.services.yml:17]: rename _user_edit_check this to _user_edit_access_check

File: modules/rte_mis_state/src/Plugin/Menu/AlterPeopleActionTask.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   */
+  public function getTitle(Request $request = NULL) {
+    // Subclasses may pull in the request or specific attributes as parameters.
+    // The title from YAML file discovery may be a TranslatableMarkup object.
+    $role = $request->query->get('role') ?? NULL;
+    if (empty($role) || !isset($role) || $role == 'All') {
+      $role = 'Add User';
+    }
+    elseif ($role == 'district_admin') {
+      $role = 'Add District User';
```
Comment (jaswinsingh) [inline modules/rte_mis_state/src/Plugin/Menu/AlterPeopleActionTask.php:73]: Add this inside `t()`

File: modules/rte_mis_state/src/Plugin/Menu/AlterPeopleActionTask.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    elseif ($role == 'block_admin') {
+      $role = 'Add Block User';
```
Comment (jaswinsingh) [inline modules/rte_mis_state/src/Plugin/Menu/AlterPeopleActionTask.php:76]: Add this inside `t()`

File: modules/rte_mis_state/src/Access/UserEditAccessCheck.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+namespace Drupal\rte_mis_state\Access;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Routing\Access\AccessInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+
+/**
+ * Determines access to the user register page.
```
Comment (jaswinsingh) [inline modules/rte_mis_state/src/Access/UserEditAccessCheck.php:12]: Update the comment here also

File: modules/rte_mis_state/rte_mis_state.module [hunk 9]
```diff
@@ -685,3 +769,56 @@ function rte_mis_state_mail($key, &$message, $params) {
...
+      $variables['content']['#title'] = t('Add District Admin');
+    }
+    else {
+      $variables['content']['#title'] = t('Add User');
+    }
+  }
+}
+
+/**
+ * Function to find the childs of a location of the User.
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:803]: Add params comment here

File: modules/rte_mis_state/rte_mis_state.module [hunk 10]
```diff
@@ -210,6 +234,31 @@ function rte_mis_state_form_alter(&$form, FormStateInterface $form_state, $form_
...
+
+    if ($role_param) {
+      if (in_array('state_admin', $curr_user_role)) {
+        $form['role']['#options'] = [
+          'All' => 'Any',
+          'school' => 'School',
+          'school_admin' => 'School Admin',
+          'block_admin' => 'Block Admin',
+          'district_admin' => 'District Admin',
+        ];
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:250]: Add this inside `t()`

File: modules/rte_mis_state/rte_mis_state.module [hunk 11]
```diff
@@ -210,6 +234,31 @@ function rte_mis_state_form_alter(&$form, FormStateInterface $form_state, $form_
...
+        $form['permission']['#access'] = FALSE;
+      }
+      elseif (in_array('district_admin', $curr_user_role)) {
+        $form['role']['#options'] = [
+          'All' => 'Any',
+          'block_admin' => 'Block Admin',
```
Comment (jaswinsingh) [inline modules/rte_mis_state/rte_mis_state.module:256]: Add this inside `t()`

**Review by jaswinsingh — APPROVED** [review]

**Pull request merged**